lib/Optimizer/Transforms/StatePreparation.cpp

```diff
@@ -0,0 +1,209 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "StateDecomposer.h"
+#include "cudaq/Optimizer/Builder/Runtime.h"
+#include "cudaq/Optimizer/CodeGen/QIRFunctionNames.h"
+#include "cudaq/Optimizer/Dialect/CC/CCOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeTypes.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Complex/IR/Complex.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Target/LLVMIR/TypeToLLVM.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/RegionUtils.h"
+#include <span>
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_STATEPREPARATION
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+#define DEBUG_TYPE "state-preparation"
+
+using namespace mlir;
+
+/// Replace a qubit initialization from vectors with quantum gates.
+/// For example:
+///
+///
+/// Before StatePreparation (state-prep):
+///
+/// module {
+///   func.func @foo() attributes {
+///     %0 = cc.address_of @foo.rodata_0 : !cc.ptr<!cc.array<complex<f32> x 4>>
+///     %1 = quake.alloca !quake.veq<2>
+///     %2 = quake.init_state %1, %0 : (!quake.veq<2>,
+///       !cc.ptr<!cc.array<complex<f32> x 4>>) -> !quake.veq<2> return
+///  }
+///  cc.global constant @foo.rodata_0 (dense<[(0.707106769,0.000000e+00),
+///      (0.707106769,0.000000e+00), (0.000000e+00,0.000000e+00),
+///      (0.000000e+00,0.000000e+00)]> : tensor<4xcomplex<f32>>) :
+///    !cc.array<complex<f32> x 4>
+/// }
```

Comment:
I'm a bit surprised this unmarked code block in the doxygen comments doesn't choke the spelling checker in the CI.

Reply:
added quotes

Reply:
Was it actually getting through? That's nothing short of amazing. heh